File: Trees/Implement_Trie.py
'''
Approach:
1. We implement the TrieNode class with a dictionary of children nodes and a flag to indicate the end of a word.
2. The Trie class is implemented with a root TrieNode as its member variable.
3. The insert function traverses through each character of the input word and creates new TrieNode objects if necessary.
4. The search function traverses through each character of the input word and returns True only if the last node is marked as the end of a word.
5. The startsWith function is similar to the search function but returns True as long as the prefix exists in the Trie.

Time Complexity:
- Insertion: O(m), where m is the length of the word being inserted.
- Search: O(m), where m is the length of the word being searched.
- Starts With: O(m), where m is the length of the prefix being checked.

Space Complexity:
- O(n*m), where n is the number of words inserted and m is the average length of the words.

Sample input and output:

trie = Trie()
trie.insert("apple")
print(trie.search("apple"))  # Output: True
print(trie.search("app"))  # Output: False
print(trie.startsWith("app"))  # Output: True
trie.insert("app")
print(trie.search("app"))  # Output: True

Here's an implementation of the Trie data structure, also known as the Prefix Tree, in Python:
'''

# Trie node class
class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_end_of_word = False


# Trie class
class Trie:
    def __init__(self):
        self.root = TrieNode()

    # Function to insert a word into the Trie
    def insert(self, word: str) -> None:
        node = self.root
        for char in word:
            # If the current character doesn't exist in the Trie, create a new node
            if char not in node.children:
                node.children[char] = TrieNode()
            # Move to next node
            node = node.children[char]
        # Mark the end of a word
        node.is_end_of_word = True

    # Function to search for a word in the Trie
    def search(self, word: str) -> bool:
        node = self.root
        for char in word:
            # If the current character doesn't exist, the word doesn't exist
            if char not in node.children:
                return False
            # Move to next node
            node = node.children[char]
        # Return True only if the last node is marked as the end of a word
        return node.is_end_of_word

    # Function to check if a word is a prefix of any existing word in the Trie
    def startsWith(self, prefix: str) -> bool:
        node = self.root
        for char in prefix:
            # If the current character doesn't exist, the prefix doesn't exist
            if char not in node.children:
                return False
            # Move to next node
            node = node.children[char]
        # The prefix exists if we reach this point
        return True

File: Trees/Implement_Trie.cpp
/*
Approach:
- We define a Trie node structure that contains an array of child nodes and a flag to indicate the end of a word.
- The Trie class provides methods to insert words into the Trie, search for words, and check if any word starts with a given prefix.
- The insert operation involves traversing the Trie for each character of the word and creating a new node if the path doesn't exist. At the end, we mark the last node as the end of a word.
- The search operation is similar to the insert operation, but it returns true only if the last node is marked as the end of a word.
- The startsWith operation also involves traversing the Trie for each character of the prefix. It returns true regardless of whether the last node is marked as the end of a word or not.

Time Complexity:
- The time complexity of inserting a word into the Trie is O(m), where m is the length of the word.
- The time complexity of searching for a word or checking for a prefix is O(m), where m is the length of the word or prefix.

Space Complexity:
- The space complexity of the Trie is O(N * M), where N is the number of words and M is the average length of the words.

Sample Input / Output:
int main() {
    Trie trie;
    
    trie.insert("apple");
    trie.insert("banana");
    
    cout << trie.search("apple") << endl;   // Output: 1 (true)
    cout << trie.search("banana") << endl;  // Output: 1 (true)
    cout << trie.search("car") << endl;     // Output: 0 (false)
    
    cout << trie.startsWith("app") << endl;     // Output: 1 (true)
    cout << trie.startsWith("ban") << endl;     // Output: 1 (true)
    cout << trie.startsWith("cab") << endl;     // Output: 0 (false)
    
    return 0;
}

*/

//Here's the code for implementing a Trie (Prefix Tree) in C++:

// Trie node structure
struct TrieNode {
    struct TrieNode* children[26];    // Array to store child nodes
    bool isEndOfWord;                 // Flag to indicate end of word
    TrieNode() {
        isEndOfWord = false;
        for (int i = 0; i < 26; i++)
            children[i] = nullptr;
    }
};

// Trie class
class Trie {
private:
    TrieNode* root;   // Pointer to the root node

public:
    Trie() {
        root = new TrieNode();
    }

    // Insert a word into the Trie
    void insert(string word) {
        TrieNode* current = root;
        
        for (char ch : word) {
            int index = ch - 'a';
            
            // Create a new node if the path doesn't exist
            if (!current->children[index]) {
                current->children[index] = new TrieNode();
            }
            
            current = current->children[index];
        }
        
        // Mark the end of a word
        current->isEndOfWord = true;
    }

    // Search for a word in the Trie
    bool search(string word) {
        TrieNode* current = root;
        
        for (char ch : word) {
            int index = ch - 'a';
            
            // Return false if the path doesn't exist
            if (!current->children[index]) {
                return false;
            }
            
            current = current->children[index];
        }
        
        // Return true only if the current node is the end of a word
        return current->isEndOfWord;
    }

    // Check if any word in the Trie starts with the given prefix
    bool startsWith(string prefix) {
        TrieNode* current = root;
        
        for (char ch : prefix) {
            int index = ch - 'a';
            
            // Return false if the path doesn't exist
            if (!current->children[index]) {
                return false;
            }
            
            current = current->children[index];
        }
        
        // Return true regardless of whether the current node is the end of a word or not
        return true;
    }
};
